File: tests/r/test-matrix.R
test_that("X nnz matches expected", {
  skip_if_no_golden(SCX_PATH)
  expect_equal(Matrix::nnzero(scx_ad()$X), EXPECTED_NNZ)
})

test_that("X nnz matches reference", {
  skip_if_no_golden(SCX_PATH, REF_PATH)
  expect_equal(Matrix::nnzero(scx_ad()$X), Matrix::nnzero(ref_ad()$X))
})

test_that("X row sums match reference", {
  skip_if_no_golden(SCX_PATH, REF_PATH)
  # X is obs x vars (cells x genes); rowSums = per-cell totals
  expect_equal(
    Matrix::rowSums(scx_ad()$X),
    Matrix::rowSums(ref_ad()$X),
    tolerance = 1e-4
  )
})

test_that("X col sums match reference", {
  skip_if_no_golden(SCX_PATH, REF_PATH)
  # colSums = per-gene totals
  expect_equal(
    Matrix::colSums(scx_ad()$X),
    Matrix::colSums(ref_ad()$X),
    tolerance = 1e-4
  )
})

test_that("spot check: first 100 obs x 100 vars exact", {
  skip_if_no_golden(SCX_PATH, REF_PATH)
  s <- as.matrix(scx_ad()$X[1:100, 1:100])
  r <- as.matrix(ref_ad()$X[1:100, 1:100])
  expect_equal(s, r, tolerance = 1e-4)
})

File: tests/python/test_matrix.py
import numpy as np
import scipy.sparse as sp
from conftest import EXPECTED_NNZ


def test_nnz(scx_adata):
    assert scx_adata.X.nnz == EXPECTED_NNZ


def test_nnz_matches_reference(scx_adata, ref_adata):
    assert scx_adata.X.nnz == ref_adata.X.nnz


def test_row_sums(scx_adata, ref_adata):
    """Total counts per cell must agree."""
    r = np.asarray(ref_adata.X.sum(axis=1)).ravel()
    s = np.asarray(scx_adata.X.sum(axis=1)).ravel()
    np.testing.assert_allclose(s, r, rtol=1e-4, err_msg="X row sums differ")


def test_col_sums(scx_adata, ref_adata):
    """Total counts per gene must agree."""
    r = np.asarray(ref_adata.X.sum(axis=0)).ravel()
    s = np.asarray(scx_adata.X.sum(axis=0)).ravel()
    np.testing.assert_allclose(s, r, rtol=1e-4, err_msg="X col sums differ")


def test_spot_values(scx_adata, ref_adata):
    """Exact value check on first 100 cells × 100 genes submatrix."""
    r = ref_adata.X[:100, :100].toarray()
    s = scx_adata.X[:100, :100].toarray()
    np.testing.assert_allclose(s, r, rtol=1e-4, err_msg="X spot values differ")


def test_x_is_sparse(scx_adata):
    assert sp.issparse(scx_adata.X), "X should be sparse"
